pkg/controller/common/beat/autodiscover.go

```diff
@@ -0,0 +1,149 @@
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/tracing"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/maps"
+)
+
+const (
+	serviceAccountNameTemplate     = "elastic-operator-autodiscover-%s"
+	clusterRoleBindingNameTemplate = "elastic-operator-autodiscover-%s-%s"
+	clusterRoleName                = "elastic-operator-beat"
+)
+
+var (
+	shouldSetupRBAC = false
+)
+
+// EnableAutodiscoverRBACSetup enables setting up autodiscover RBAC.
+func EnableAutodiscoverRBACSetup() {
+	shouldSetupRBAC = true
+}
+
+// ShouldSetupAutodiscoverRBAC returns true if autodiscover RBAC is expected to be set up by the operator.
+func ShouldSetupAutodiscoverRBAC() bool {
+	return shouldSetupRBAC
+}
+
+// SetupAutodiscoveryRBAC reconciles all resources needed for the default RBAC setup.
+func SetupAutodiscoverRBAC(ctx context.Context, log logr.Logger, client k8s.Client, owner metav1.Object, labels map[string]string) error {
+	if ShouldSetupAutodiscoverRBAC() {
+		if err := setupAutodiscoverRBAC(ctx, client, owner, labels); err != nil {
+			log.V(1).Info(
+				"autodiscovery rbac setup failed",
+				"namespace", owner.GetNamespace(),
+				"beat_name", owner.GetName())
+			return err
+		}
+	}
+	return nil
+}
+
+func setupAutodiscoverRBAC(ctx context.Context, client k8s.Client, owner metav1.Object, labels map[string]string) error {
+	// this is the source of truth and should be respected at all times
+	if !ShouldSetupAutodiscoverRBAC() {
+		return nil
+	}
+
+	span, _ := apm.StartSpan(ctx, "reconcile_autodiscover_rbac", tracing.SpanTypeApp)
+	defer span.End()
+
+	if err := reconcileServiceAccount(client, owner, labels); err != nil {
+		return err
+	}
+
+	if err := reconcileClusterRoleBinding(client, owner); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func reconcileServiceAccount(client k8s.Client, owner metav1.Object, labels map[string]string) error {
+	sa := &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      ServiceAccountName(owner.GetName()),
+			Namespace: owner.GetNamespace(),
+			Labels:    labels,
+		},
+	}
+
+	return doReconcile(client, sa, owner)
+}
+
+func reconcileClusterRoleBinding(client k8s.Client, owner metav1.Object) error {
+	binding := &rbacv1.ClusterRoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: ClusterRoleBindingName(owner.GetNamespace(), owner.GetName()),
+		},
+		Subjects: []rbacv1.Subject{
+			{
+				Kind:      "ServiceAccount",
+				Name:      ServiceAccountName(owner.GetName()),
+				Namespace: owner.GetNamespace(),
+			},
+		},
+		RoleRef: rbacv1.RoleRef{
+			APIGroup: rbacv1.GroupName,
+			Kind:     "ClusterRole",
+			Name:     clusterRoleName,
+		},
+	}
+
+	return doReconcile(client, binding, nil)
+}
+
+func doReconcile(client k8s.Client, obj runtime.Object, owner metav1.Object) error {
+	// labels set here must be exactly the same for all callers in a particular namespace
+	// otherwise they'll just keep trying to override each other
```

Comment:
Since one of the resources (clusterrole) reconciled is shared between many Beats, it can't have different labels depending on which Beat resource is currently being reconciled. This is no longer needed after recent changes.